File: android/core/src/main/kotlin/com/iptvcar/core/provider/xtream/XtreamParser.kt
package com.iptvcar.core.provider.xtream

import com.iptvcar.core.model.Category
import com.iptvcar.core.model.ContentType
import com.iptvcar.core.model.Episode
import com.iptvcar.core.model.LiveChannel
import com.iptvcar.core.model.Movie
import com.iptvcar.core.model.Season
import com.iptvcar.core.model.Series
import org.json.JSONArray
import org.json.JSONObject

/**
 * Parses raw Xtream Codes `player_api.php` JSON per shared/parsing-spec/xtream-and-m3u.md.
 * Every accessor is lenient: a missing or mistyped field degrades to null/default
 * rather than throwing, because Xtream server implementations vary widely.
 */
object XtreamParser {

    private fun JSONObject.optStringOrNull(key: String): String? =
        if (has(key) && !isNull(key)) getString(key) else null

    private fun JSONObject.optIntOrNull(key: String): Int? {
        if (!has(key) || isNull(key)) return null
        return when (val v = get(key)) {
            is Int -> v
            is Number -> v.toInt()
            is String -> v.toIntOrNull()
            else -> null
        }
    }

    private fun JSONObject.optDoubleOrNull(key: String): Double? {
        if (!has(key) || isNull(key)) return null
        return when (val v = get(key)) {
            is Number -> v.toDouble()
            is String -> v.toDoubleOrNull()
            else -> null
        }
    }

    fun parseCategories(providerId: String, contentType: ContentType, json: String): List<Category> {
        val array = JSONArray(json)
        return (0 until array.length()).map { i ->
            val o = array.getJSONObject(i)
            Category(
                id = o.getString("category_id"),
                providerId = providerId,
                name = o.optString("category_name", "Unnamed"),
                contentType = contentType,
            )
        }
    }

    fun parseLiveStreams(
        providerId: String,
        credentials: XtreamCredentials,
        json: String,
    ): List<LiveChannel> {
        val array = JSONArray(json)
        return (0 until array.length()).map { i ->
            val o = array.getJSONObject(i)
            val streamId = o.get("stream_id").toString()
            LiveChannel(
                id = streamId,
                providerId = providerId,
                categoryId = o.optStringOrNull("category_id") ?: "uncategorized",
                name = o.optString("name", "Unnamed channel"),
                logoUrl = o.optStringOrNull("stream_icon"),
                streamUrl = liveStreamUrl(credentials, streamId, extension = "m3u8"),
                epgChannelId = o.optStringOrNull("epg_channel_id"),
                catchupAvailable = o.optIntOrNull("tv_archive") == 1,
                catchupDays = o.optIntOrNull("tv_archive_duration"),
            )
        }
    }

    fun parseVodStreams(
        providerId: String,
        credentials: XtreamCredentials,
        json: String,
    ): List<Movie> {
        val array = JSONArray(json)
        return (0 until array.length()).map { i ->
            val o = array.getJSONObject(i)
            val streamId = o.get("stream_id").toString()
            val ext = o.optStringOrNull("container_extension") ?: "mp4"
            Movie(
                id = streamId,
                providerId = providerId,
                categoryId = o.optStringOrNull("category_id") ?: "uncategorized",
                title = o.optString("name", "Untitled"),
                posterUrl = o.optStringOrNull("stream_icon"),
                rating = o.optDoubleOrNull("rating"),
                streamUrl = vodStreamUrl(credentials, streamId, ext),
                containerExtension = ext,
            )
        }
    }

    /** Enriches a Movie with detail from `get_vod_info`. */
    fun parseVodInfo(movie: Movie, json: String): Movie {
        val root = JSONObject(json)
        val info = root.optJSONObject("info") ?: return movie
        return movie.copy(
            description = info.optStringOrNull("plot") ?: movie.description,
            genre = info.optStringOrNull("genre") ?: movie.genre,
            durationSeconds = info.optIntOrNull("duration_secs") ?: movie.durationSeconds,
            year = info.optStringOrNull("releasedate")?.take(4)?.toIntOrNull() ?: movie.year,
        )
    }

    fun parseSeriesList(providerId: String, json: String): List<Series> {
        val array = JSONArray(json)
        return (0 until array.length()).map { i ->
            val o = array.getJSONObject(i)
            Series(
                id = o.get("series_id").toString(),
                providerId = providerId,
                categoryId = o.optStringOrNull("category_id") ?: "uncategorized",
                title = o.optString("name", "Untitled"),
                posterUrl = o.optStringOrNull("cover"),
                rating = o.optDoubleOrNull("rating"),
            )
        }
    }

    /** Parses `get_series_info` into full Season/Episode trees for one series. */
    fun parseSeriesInfo(series: Series, credentials: XtreamCredentials, json: String): Series {
        val root = JSONObject(json)
        val info = root.optJSONObject("info")
        val enriched = if (info != null) {
            series.copy(
                description = info.optStringOrNull("plot") ?: series.description,
                genre = info.optStringOrNull("genre") ?: series.genre,
                posterUrl = info.optStringOrNull("cover") ?: series.posterUrl,
            )
        } else series

        val episodesByseason = root.optJSONObject("episodes") ?: JSONObject()
        val seasons = episodesByseason.keys().asSequence().map { seasonKey ->
            val seasonNumber = seasonKey.toIntOrNull() ?: 0
            val episodesArray = episodesByseason.getJSONArray(seasonKey)
            val episodes = (0 until episodesArray.length()).map { i ->
                val eo = episodesArray.getJSONObject(i)
                val epInfo = eo.optJSONObject("info")
                val ext = eo.optStringOrNull("container_extension") ?: "mp4"
                val episodeId = eo.get("id").toString()
                Episode(
                    id = episodeId,
                    seasonId = "${series.id}-s$seasonNumber",
                    episodeNumber = eo.optIntOrNull("episode_num") ?: 0,
                    title = eo.optString("title", "Episode"),
                    durationSeconds = epInfo?.optIntOrNull("duration_secs"),
                    description = epInfo?.optStringOrNull("plot"),
                    streamUrl = episodeStreamUrl(credentials, episodeId, ext),
                    containerExtension = ext,
                )
            }
            Season(
                id = "${series.id}-s$seasonNumber",
                seriesId = series.id,
                seasonNumber = seasonNumber,
                episodes = episodes,
            )
        }.sortedBy { it.seasonNumber }.toList()

        return enriched.copy(seasons = seasons)
    }

    fun liveStreamUrl(c: XtreamCredentials, streamId: String, extension: String) =
        "${c.baseUrl.trimEnd('/')}/live/${c.username}/${c.password}/$streamId.$extension"

    fun vodStreamUrl(c: XtreamCredentials, streamId: String, extension: String) =
        "${c.baseUrl.trimEnd('/')}/movie/${c.username}/${c.password}/$streamId.$extension"

    fun episodeStreamUrl(c: XtreamCredentials, episodeId: String, extension: String) =
        "${c.baseUrl.trimEnd('/')}/series/${c.username}/${c.password}/$episodeId.$extension"

    /** Returns true only when the auth JSON positively confirms a valid session. */
    fun isAuthenticated(json: String): Boolean {
        return try {
            val root = JSONObject(json)
            val userInfo = root.optJSONObject("user_info") ?: return false
            userInfo.optString("auth", "0") == "1"
        } catch (e: Exception) {
            false
        }
    }
}

File: ios/IPTVCar/Core/Parsing/XtreamParser.swift
import Foundation

/// Parses raw Xtream Codes `player_api.php` JSON per
/// shared/parsing-spec/xtream-and-m3u.md. Mirrors
/// android/core/.../xtream/XtreamParser.kt field-for-field. Every accessor
/// is lenient: a missing or mistyped field degrades to nil rather than
/// throwing, because Xtream server implementations vary widely.
enum XtreamParser {

    private static func string(_ any: Any?) -> String? {
        if let s = any as? String { return s }
        if let n = any as? NSNumber { return n.stringValue }
        return nil
    }

    private static func int(_ any: Any?) -> Int? {
        if let n = any as? NSNumber { return n.intValue }
        if let s = any as? String { return Int(s) }
        return nil
    }

    private static func double(_ any: Any?) -> Double? {
        if let n = any as? NSNumber { return n.doubleValue }
        if let s = any as? String { return Double(s) }
        return nil
    }

    static func isAuthenticated(_ data: Data) -> Bool {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let userInfo = root["user_info"] as? [String: Any] else { return false }
        return string(userInfo["auth"]) == "1"
    }

    static func parseCategories(providerId: String, contentType: ContentType, data: Data) throws -> [Category] {
        guard let array = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            throw XtreamError.malformedResponse("Expected a JSON array of categories")
        }
        return array.map { o in
            Category(
                id: string(o["category_id"]) ?? UUID().uuidString,
                providerId: providerId,
                name: string(o["category_name"]) ?? "Unnamed",
                contentType: contentType,
                parentId: nil
            )
        }
    }

    static func liveStreamUrl(_ c: XtreamCredentials, streamId: String, ext: String = "m3u8") -> String {
        "\(c.baseUrl)/live/\(c.username)/\(c.password)/\(streamId).\(ext)"
    }

    static func vodStreamUrl(_ c: XtreamCredentials, streamId: String, ext: String) -> String {
        "\(c.baseUrl)/movie/\(c.username)/\(c.password)/\(streamId).\(ext)"
    }

    static func episodeStreamUrl(_ c: XtreamCredentials, episodeId: String, ext: String) -> String {
        "\(c.baseUrl)/series/\(c.username)/\(c.password)/\(episodeId).\(ext)"
    }

    static func parseLiveStreams(providerId: String, credentials: XtreamCredentials, data: Data) throws -> [LiveChannel] {
        guard let array = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            throw XtreamError.malformedResponse("Expected a JSON array of live streams")
        }
        return array.map { o in
            let streamId = string(o["stream_id"]) ?? UUID().uuidString
            return LiveChannel(
                id: streamId,
                providerId: providerId,
                categoryId: string(o["category_id"]) ?? "uncategorized",
                name: string(o["name"]) ?? "Unnamed channel",
                logoUrl: string(o["stream_icon"]),
                streamUrl: liveStreamUrl(credentials, streamId: streamId),
                epgChannelId: string(o["epg_channel_id"]),
                catchupAvailable: int(o["tv_archive"]) == 1,
                catchupDays: int(o["tv_archive_duration"])
            )
        }
    }

    static func parseVodStreams(providerId: String, credentials: XtreamCredentials, data: Data) throws -> [Movie] {
        guard let array = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            throw XtreamError.malformedResponse("Expected a JSON array of VOD streams")
        }
        return array.map { o in
            let streamId = string(o["stream_id"]) ?? UUID().uuidString
            let ext = string(o["container_extension"]) ?? "mp4"
            return Movie(
                id: streamId,
                providerId: providerId,
                categoryId: string(o["category_id"]) ?? "uncategorized",
                title: string(o["name"]) ?? "Untitled",
                year: nil,
                posterUrl: string(o["stream_icon"]),
                durationSeconds: nil,
                description: nil,
                genre: nil,
                rating: double(o["rating"]),
                streamUrl: vodStreamUrl(credentials, streamId: streamId, ext: ext),
                containerExtension: ext
            )
        }
    }

    static func parseVodInfo(movie: Movie, data: Data) -> Movie {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let info = root["info"] as? [String: Any] else { return movie }
        var updated = movie
        updated.description = string(info["plot"]) ?? movie.description
        updated.genre = string(info["genre"]) ?? movie.genre
        updated.durationSeconds = int(info["duration_secs"]) ?? movie.durationSeconds
        if let release = string(info["releasedate"]), release.count >= 4 {
            updated.year = Int(release.prefix(4)) ?? movie.year
        }
        return updated
    }

    static func parseSeriesList(providerId: String, data: Data) throws -> [Series] {
        guard let array = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            throw XtreamError.malformedResponse("Expected a JSON array of series")
        }
        return array.map { o in
            Series(
                id: string(o["series_id"]) ?? UUID().uuidString,
                providerId: providerId,
                categoryId: string(o["category_id"]) ?? "uncategorized",
                title: string(o["name"]) ?? "Untitled",
                posterUrl: string(o["cover"]),
                description: nil,
                genre: nil,
                rating: double(o["rating"]),
                seasons: []
            )
        }
    }

    static func parseSeriesInfo(series: Series, credentials: XtreamCredentials, data: Data) -> Series {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return series }
        var updated = series
        if let info = root["info"] as? [String: Any] {
            updated.description = string(info["plot"]) ?? series.description
            updated.genre = string(info["genre"]) ?? series.genre
            updated.posterUrl = string(info["cover"]) ?? series.posterUrl
        }

        guard let episodesBySeason = root["episodes"] as? [String: [[String: Any]]] else { return updated }

        let seasons: [Season] = episodesBySeason.keys.compactMap { seasonKey in
            let seasonNumber = Int(seasonKey) ?? 0
            let episodesArray = episodesBySeason[seasonKey] ?? []
            let episodes: [Episode] = episodesArray.map { eo in
                let epInfo = eo["info"] as? [String: Any]
                let ext = string(eo["container_extension"]) ?? "mp4"
                let episodeId = string(eo["id"]) ?? UUID().uuidString
                return Episode(
                    id: episodeId,
                    seasonId: "\(series.id)-s\(seasonNumber)",
                    episodeNumber: int(eo["episode_num"]) ?? 0,
                    title: string(eo["title"]) ?? "Episode",
                    posterUrl: nil,
                    durationSeconds: epInfo.flatMap { int($0["duration_secs"]) },
                    description: epInfo.flatMap { string($0["plot"]) },
                    streamUrl: episodeStreamUrl(credentials, episodeId: episodeId, ext: ext),
                    containerExtension: ext
                )
            }
            return Season(id: "\(series.id)-s\(seasonNumber)", seriesId: series.id, seasonNumber: seasonNumber, episodes: episodes.sorted { $0.episodeNumber < $1.episodeNumber })
        }.sorted { $0.seasonNumber < $1.seasonNumber }

        updated.seasons = seasons
        return updated
    }
}
